Firestore/core/test/firebase/firestore/local/leveldb_query_cache_test.cc

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/include/firebase/firestore/timestamp.h"
+#include "Firestore/core/src/firebase/firestore/local/leveldb_persistence.h"
+#include "Firestore/core/src/firebase/firestore/local/leveldb_query_cache.h"
+#include "Firestore/core/src/firebase/firestore/local/persistence.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+#include "Firestore/core/test/firebase/firestore/local/persistence_testing.h"
+#include "Firestore/core/test/firebase/firestore/local/query_cache_test.h"
+#include "Firestore/core/test/firebase/firestore/testutil/testutil.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+namespace {
+
+using model::DocumentKey;
+using model::ListenSequenceNumber;
+using model::SnapshotVersion;
+using model::TargetId;
+using util::Path;
+using testutil::Query;
+
+std::unique_ptr<Persistence> PersistenceFactory() {
+  return LevelDbPersistenceForTesting();
+}
+
+}  // namespace
+
+INSTANTIATE_TEST_CASE_P(LevelDbQueryCacheTest,
+                        QueryCacheTest,
+                        testing::Values(PersistenceFactory));
+
+class LevelDbQueryCacheTest : public QueryCacheTestBase {
+ public:
+  LevelDbQueryCacheTest()
+      : QueryCacheTestBase(PersistenceFactory()) {
+  }
+
+  LevelDbQueryCache* leveldb_cache() {
+    return static_cast<LevelDbQueryCache*>(persistence_->query_cache());
+  }
+};
+
+TEST_F(LevelDbQueryCacheTest, MetadataPersistedAcrossRestarts) {
+  persistence_->Shutdown();
+  persistence_.reset();
+
+  Path dir = LevelDbDir();
+
+  auto db1 = LevelDbPersistenceForTesting(dir);
+  LevelDbQueryCache* query_cache = db1->query_cache();
+
+  ASSERT_EQ(0, query_cache->highest_listen_sequence_number());
+  ASSERT_EQ(0, query_cache->highest_target_id());
+  SnapshotVersion version_zero;
+  ASSERT_EQ(version_zero, query_cache->GetLastRemoteSnapshotVersion());
+
+  ListenSequenceNumber minimum_sequence_number = 1234;
+  TargetId last_target_id = 5;
+  SnapshotVersion last_version(Timestamp(1, 2));
+
+  db1->Run("add query data", [&] {
+    core::Query query = Query("some/path");
```

Comment:
`using core::Query` and change this line to

`Query query("some/path")`?

Reply:
These are actually different things: it's `core::Query` being assigned from the result of invoking the `testutil::Query` function.

I've updated this to make it `using core::Query` and showing `testutil::Query` on the RHS.